Analyze - First Preprocessing

liermannj opened the pull request:
Add Preprocessing module to compute features for audio signals

Currently it does :
- extract features from Fourier and MFC transform
- aggregate features from multiple unit signal into an Enreg FootPrint

Commit:
feat - analyze: first try for feature extractions

- add some general signal processing
- add fouriers coefs
- add mfc

Commit:
feat - analyze: add preparator env

- preparator is entry point for data transformation (before truely analyze)
- add private fileInput env for batch reading (demo)
- begin to add configuration, and refacto other envs to take into account env instead of raw values
- in addition, add package private annotaion over type alias for jobEnvironment

Commit:
feat - analyze: add Partial Job process for demo

- add output service
- add Job object

Commit:
feat - analyze: add partial footprint computation

- based on come fourier + mfc coefs at the moment
- does not do any analyzing properly yet, only data transformation
- add Matrix concept for vertical aggregations
- add error handling with Either and Seqs

Commit:
improve - analyze: add UT and IT

- did some refacto for more injection in function, to make tests easier
- did temporary global IT

fpatinCarrefour commented on analyze/src/main/scala/com/jliermann/analyze/RootConfiguration.scala:
```diff
@@ -0,0 +1,41 @@
+
+  implicit val charSeqReader: ConfigReader[Seq[Char]] = ConfigReader[Seq[String]].map(_.map(_.head))
+
+  def loadConfigOrThrow(config: Config): RootConfiguration = {
```
comment ton job utilise `Either` tu peux utiliser la méthode de pureconfig qui retourne `Either`

fpatinCarrefour commented on analyze/src/main/scala/com/jliermann/analyze/Job.scala:
```diff
@@ -0,0 +1,30 @@
+    val result = for {
+      input: Seq[Seq[Signal]] <- env.fileInput.readEnregs(env, localConfiguration.input)
+      coefs = input.map(env.preparator.prepareEnreg(env, config.analyzeConfiguration))
+      stringCoefs: Seq[String] = coefs
+        .rights
+        .map(_.mkString(";"))
```
c'est surprenant d'ignorer les `Left`.
de plus je ne vois pas l’intérêt de ce morceau de code ici.
Enfin fais un choix entre `Try` et `Either`, cela simplifiera ton code je pense

fpatinCarrefour commented on analyze/src/main/scala/com/jliermann/analyze/io/FileInputLive.scala:
```diff
@@ -0,0 +1,37 @@
+    for {
+      file: Seq[String] <- env.fileInput.readFromFile(config.file, config.codec)
+      sampledLines: Seq[Seq[String]] <- Try(file.map(_.split(config.sampleSep).toSeq))
+      numberedFile <- Try {
+        sampledLines
+          .map(_.map(_.filterNot(config.ignoredChars.contains)))
+          .map(_.map(tryToDouble(config.numberSep)))
+          .map(_.filterNot(_.exists(_.isFailure)))
+          .map(_.map(_.collect { case Success(d) => d }))
+      }
```
c'est assez moche, mais très bien aligné ;)

fpatinCarrefour commented on analyze/src/main/scala/com/jliermann/analyze/io/FileOutputLive.scala:
```diff
@@ -0,0 +1,19 @@
+
+private[analyze] trait FileOutputLive extends FileOutput.Service {
+
+  override def writeToFile(file: File, data: Seq[String]): Try[Unit] = Try {
+    val writer = new PrintWriter(file)
+    data
+      .map(s => s"$s\n")
+      .foreach(writer.write)
+
+    writer.close()
+  }
```
Pas de gestion de ressource.
sauf si tu as un besoin spécifique de PrintWriter, la gestion des erreurs à un côté Visual Basic
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/PrintWriter.html

Commit:
improve - analyze: improve error handling + file writing

- replace either with Seq[Try]
- enhance implicit functions over Seq[Try] to handle the different cases
- add a composite error type to allow squashing
- use BuffredOutputStream for fileWriting

jeliermann committed:
fix - sound: increase mic sensibility

- hardware changed, mic config changed, life is life

jeliermann committed:
improved - analyze : simplify inputing

- remove ignoring char, previous step should be cleaner

jeliermann committed:
feat - analyze : correct mfc computation

- add melScale conversion using natural logarithm
- add some comments, just to get reused to the code

jeliermann committed:
improve - analysis: simplify configuration

- reduce level of nesting in config
- re-add normalization at the end of the process

jeliermann committed:
fix - analyze: make test OK again

jeliermann committed:
improve: add README for analyze

jeliermann committed:
feat - analyze: conserve fourier and mfc in a sparse space

- features are now a 1024 vector
- preservation of voice timber as fundamentals are now encoded in
  different features depending on how crying puppy they sound like
- mfc are now sparse too
- sparsity is achieved in the same way as fourier, with triangular
  overlapping window

jeliermann committed:
feat - analyze: enable temporal series

- aggregation is now optional
- if aggregate is false, aggregation is replaced by temporal frames
  separated by a different tag
